From f4f465d7868bced863fa4c4711af76ff41f84ea6 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:15:18 -0400
Subject: [net] gre: Add support for checksum unnecessary conversions

Message-id: <1409652920-431-56-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90167
O-Subject: [PATCH 55/57] [RHEL7.1 net] gre: Add support for checksum unnecessary conversions
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit 884d338c041c2aa4536ade8620efa585e7c57f3c
Author: Tom Herbert <therbert@google.com>
Date:   Sun Aug 31 15:12:44 2014 -0700

    gre: Add support for checksum unnecessary conversions

    Call skb_checksum_try_convert and skb_gro_checksum_try_convert
    after checksum is found present and validated in the GRE header
    for normal and GRO paths respectively.

    In GRO path, call skb_gro_checksum_try_convert

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/gre_demux.c b/net/ipv4/gre_demux.c
index e25b80e..44e7918 100644
--- a/net/ipv4/gre_demux.c
+++ b/net/ipv4/gre_demux.c
@@ -124,6 +124,10 @@ static int parse_gre_header(struct sk_buff *skb, struct tnl_ptk_info *tpi,
    *csum_err = true;
    return -EINVAL;
   }
+
+  skb_checksum_try_convert(skb, IPPROTO_GRE, 0,
+      null_compute_pseudo);
+
   options++;
  }
 
diff --git a/net/ipv4/gre_offload.c b/net/ipv4/gre_offload.c
index f4e4eec..0b94889 100644
--- a/net/ipv4/gre_offload.c
+++ b/net/ipv4/gre_offload.c
@@ -172,10 +172,14 @@ static struct sk_buff **gre_gro_receive(struct sk_buff **head,
  }
 
  /* Don't bother verifying checksum if we're going to flush anyway. */
- if ((greh->flags & GRE_CSUM) && !NAPI_GRO_CB(skb)->flush &&
-     skb_gro_checksum_simple_validate(skb))
+ if ((greh->flags & GRE_CSUM) && !NAPI_GRO_CB(skb)->flush) {
+  if (skb_gro_checksum_simple_validate(skb))
    goto out_unlock;
 
+  skb_gro_checksum_try_convert(skb, IPPROTO_GRE, 0,
+          null_compute_pseudo);
+ }
+
  flush = 0;
 
  for (p = *head; p; p = p->next) {
-- 
1.7.1